proposals/0118-partitioned-epoch-reward-distribution.md

```diff
@@ -0,0 +1,265 @@
+
+   1. rewards distribution phase - during which the calculated epoch rewards
+      for the active stake accounts are distributed.
+
+A new sysvar account, `EpochRewards`, will track and verify the reward
+distribution during rewarding phases.
+
+
+### Rewards Calculation
+
+The reward calculation will be performed at the first block of the epoch, block
+height `X`. (Some documentation on how rewards are calculated is available
+[here](https://docs.solanalabs.com/consensus/stake-delegation-and-rewards#basics).)
+Once the full rewards are calculated, the rewards will be partitioned into
+distribution chunks stored in the bank, which will then be distributed during
+the `reward distribution` phase.
+
+To ensure that each block distributes a subset of the rewards in a
+deterministic manner for the current epoch, while also randomizing the
+distribution across different epochs, the partitioning of all rewards will be
+done as follows.
+
+To minimize the impact on block processing time during the reward distribution
+phase, a target of 4,096 stake rewards will be distributed per block. The total
+number of blocks `M` needed to distributed rewards is given by the following
+formula to round up to the nearest integer without using floating point value
+arithmetic:
+
+```
+M = ((4096 - 1)+num_stake_accounts)/4096
+```
+
+To safeguard against the number of stake accounts growing dramatically and
+overflowing the number of blocks in an epoch, the number of blocks is capped at
+10% of the total slots in an epoch. If the block cap is reached, accounts per
+partition is allowed to surpass the 4,096 target.
+
+The [SipHash 1-3](https://www.aumasson.jp/siphash/siphash.pdf) pseudo-random
+function is used to hash stake account addresses efficiently and uniformly
+across the blocks in the reward distribution phase. The hashing function for an
+epoch is created by seeding a new `SipHasher` with the parent block's blockhash.
+This hashing function can then be used to hash each active stake account's
+address into a u64 hash value. The reward distribution block index `I` can then
+be computed by applying the following formula to the hash:
+
+```
+I = (M * stake_address_hash) / 2^64
+```
+
+### Reward Distribution
+
+The reward distribution phase happens after the reward computation phase, ie.
+during the second block in the epoch as per the current design. It lasts for `M`
+blocks (see previous section).
+
+#### `EpochRewards` Sysvar Account
+
+The `EpochRewards` sysvar account records whether the rewards distribution phase
+is in progress, as well as the details needed to resume distribution when
+starting from a snapshot during the reward distribution phase. These details
+include: the parameters needed to recalculate the reward partitions, the total
+rewards to be distributed, and the amount of rewards distributed so far.
+
+The layout of `EpochRewards` sysvar is shown in the following pseudo code.
```

Comment:
The alignment is somewhat unspecified here. In repr(C), u128 either has 8 or 16 byte alignment depending on architecture. AFAIK bool has 8 byte alignment 🤔

I suggest putting offsets like so `/* 0x08 */` as prefixes to make it unambiguous. 

The padding would obviously be zeros, but it wouldn't hurt to mention here.

Reply:
We do specify 8-byte alignment in the comment on L152, but certainly doesn't hurt to add offsets, and I'll remove the "ie".
Hopefully that u128 won't be too much of a PITA. If I understand [this issue](https://github.com/rust-lang/rust/issues/54341) correctly, rust will always align u128 to 8 bytes even with repr(C), but C programs may not.

Reply:
move `active` to the end?

Reply:
Ah, thanks for this, @ripatel-fd . Seems to contradict some of what I read about this.
But anyway, I'll do some reworking and update this on Monday.

Reply:
Updated to 16-byte alignment, including moving `active` to the end